### PR TITLE
Port Jax AI stack minigpt example

### PR DESCRIPTION
**Pull request by samanklesaria**
This PR copies the Jax AI stack example https://docs.jaxstack.ai/en/latest/JAX_for_LLM_pretraining.html to the flax examples directory unchanged. The nnx apis used still seem to be appropriate, so no adjustments are required.

### CODE REVIEW

File: docs_nnx/examples/minigpt.md [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+- `devices`: This will take the value of [`jax.experimental.mesh_utils((4, 2))`](https://jax.readthedocs.io/en/latest/jax.experimental.mesh_utils.html), enabling us to build a device mesh. It is a NumPy ndarray with JAX devices (a list of devices from the JAX backend as obtained from [`jax.devices()`](https://jax.readthedocs.io/en/latest/_autosummary/jax.devices.html#jax.devices))..
+- `axis_names`, where:
+  - `batch`: 4 devices along the first axis - i.e. sharded into 4 - for data parallelism; and
+  - `model`: 2 devices along the second axis - i.e. sharded into 2 -  for tensor parallism
+
+Let's instantiate `Mesh` as `mesh` and declare the TPU configuration to define how data and model parameters are distributed across the devices:
+<!-- #endregion -->
+
+```python
+mesh = Mesh(mesh_utils.create_device_mesh((4, 2)), ("batch", "model"))
```
Comment (vfdev-5) [inline docs_nnx/examples/minigpt.md]: I think we can use jax.make_mesh here instead of experimental jax feature

File: docs_nnx/examples/minigpt.md [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+                                  kernel_init=nnx.with_partitioning(nnx.initializers.xavier_uniform(), NamedSharding(mesh, P(None, 'model'))),
+                                  bias_init=nnx.with_partitioning(nnx.initializers.zeros_init(), NamedSharding(mesh, P('model'))),
+                                  rngs=rngs)
+        # The second dropout with `flax.nnx.Dropout`.
+        self.dropout2 = nnx.Dropout(rate=rate)
+        # Second layer normalization with `flax.nnx.LayerNorm`.
+        self.layer_norm2 = nnx.LayerNorm(epsilon=1e-6,
+                                         num_features=embed_dim,
+                                         scale_init=nnx.with_partitioning(nnx.initializers.ones_init(), NamedSharding(mesh, P(None, 'model'))),
+                                         bias_init=nnx.with_partitioning(nnx.initializers.zeros_init(), NamedSharding(mesh, P(None, 'model'))),
```
Comment (vfdev-5) [inline docs_nnx/examples/minigpt.md]: Let's use scale/bias metadat instead of overriding the initializers

File: docs_nnx/examples/minigpt.md [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+%load_ext tensorboard
+```
+
+As we're going to be running this model a number of times, we need some scaffolding to more easily compare our work. For a baseline, we'll need to perform some warmup to guarantee that our code is JIT'd and that our TPUs are warm. For improved comparability, we'll only start tracing after we've finished warmup.
+
+```python
+trace_dir = "/tmp/jax-trace/"
+
+def loop_step(batch, step):
+    input_batch = jnp.array(jnp.array(batch).T)
```
Comment (vfdev-5) [inline docs_nnx/examples/minigpt.md]: Do we need to wrap it twice?

File: docs_nnx/examples/minigpt.md [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+
+```python
+# Defines the loss function using `optax.softmax_cross_entropy_with_integer_labels`.
+def loss_fn(model, batch):
+    logits = model(batch[0])
+    loss = optax.softmax_cross_entropy_with_integer_labels(logits=logits, labels=batch[1]).mean()
+    return loss, logits
+
+# Define the training step with the `flax.nnx.jit` transformation decorator.
+@nnx.jit
```
Comment (vfdev-5) [inline docs_nnx/examples/minigpt.md]: let's donate model and optimizer args

File: docs_nnx/examples/minigpt.md (hunk 1 above)
Comment (samanklesaria) [reply docs_nnx/examples/minigpt.md]: Good catch. Fixed.

File: docs_nnx/examples/minigpt.md (hunk 2 above)
Comment (samanklesaria) [reply docs_nnx/examples/minigpt.md]: I didn't know about those arguments! Fixed.

File: docs_nnx/examples/minigpt.md (hunk 4 above)
Comment (samanklesaria) [reply docs_nnx/examples/minigpt.md]: Done, I think.

**Comment by samanklesaria** [comment]
Actually, it seems we need to do a little more to make this work with the current NNX. Currently getting some sharding errors. Will debug.

File: docs_nnx/examples/minigpt.md (hunk 3 above)
Comment (samanklesaria) [reply docs_nnx/examples/minigpt.md]: Good catch. Fixed.

**Comment by samanklesaria** [comment]
Sharding issues resolved. Have to wait until I have access to multiple gpus to run the full notebook though.

**Comment by samanklesaria** [comment]
I also exposed the `is_causal` argument from the `dot_product_attention` function to `MultiHeadAttention`. It seems like the user is otherwise expected to generate the causal mask themselves.

**Comment by cgarciae** [comment]
Looks good but there are some merge issues.

**Review by cgarciae — APPROVED** [review]
